File: src/www/__test__/building/Post_20221019_Farm_Context.js
import { getByTestId, screen } from "@testing-library/react";
import { mainView } from "../main";
import { waitForEnterTheGame, waitForReloadGame } from "../game/actions";
import { waitForEndMoon } from "../moon/actions";

export class Post_20221019_Farm_Context {
  async beforeTest() {
    // Do your setup here
    await waitForEnterTheGame();
  }

  async thereShouldBeNSCard(expected, s1) {
    // text:  * there should be 0 "Farm" card.
    // code: await this.thereShouldBeNSCard(0, "Farm")
    // hint: Post_20221020_FriendlyWolf_Context.thereShouldBeNSCard

    var actual = expected; // FIXME
    expect(actual).toEqual(expected);

  }

  async givenThereIsNStackOfNSAtLevelNAndNXpNSNSNSCards(
    n1,
    n2,
    s1,
    n3,
    n4,
    n5,
    s2,
    n6,
    s3,
    n7,
    s4
  ) {
    // text:  * given there is 1 stack of 1 "Build Idea" at level 2 and 6 XP, 1 "Villager", 1 "Stone", 2 "Wood" cards
    // code: await this.givenThereIsNStackOfNSAtLevelNAndNXpNSNSNSCards(1, 1, "Build Idea", 2, 6, 1, "Villager", 1, "Stone", 2, "Wood")

    await waitForReloadGame();


  }

  async endTheCurrentMoon() {
    // text:  * end the current moon.
    // code: await this.endTheCurrentMoon()
    // hint: Post_20221024_Pickaxe_Context.endTheCurrentMoon

    await waitForEndMoon();

  }

  async thereShouldBeNStackOfNSNSAndNSCards(expected, n2, s1, n3, s2, n4, s3) {
    // text:  * there should be 1 stack of 0 "Stone", 0 "Wood" and 0 "Berry Bush" cards.
    // code: await this.thereShouldBeNStackOfNSNSAndNSCards(1, 0, "Stone", 0, "Wood", 0, "Berry Bush")
    // hint: Post_20221024_Diamond_Context.thereShouldBeNStacksOfNSNSAndNSCards

    var actual = expected; // FIXME
    expect(actual).toEqual(expected);

  }

  async givenANewEmptyGame() {
    // text:  * Given a new empty game
    // code: await this.givenANewEmptyGame()

    await waitForReloadGame();

  }

  async givenThereIsNStackOfNSCardsNSCardsAndNSCardsOrAnotherPlant(
    n1,
    n2,
    s1,
    n3,
    s2,
    n4,
    s3
  ) {
    // text:  * Given there is 1 stack of 1 "Farm" cards, 1 "Villager" cards and 1 "Berry Bush" cards or another plant.
    // code: await this.givenThereIsNStackOfNSCardsNSCardsAndNSCardsOrAnotherPlant(1, 1, "Farm", 1, "Villager", 1, "Berry Bush")

    await waitForReloadGame();

  }

  async thereShouldBeNStackOfNSCardsNSCardsNSAndNSCards(
    expected,
    n2,
    s1,
    n3,
    s2,
    n4,
    s3,
    n5,
    s4
  ) {
    // text:  * There should be 1 stack of 1 "Farm" cards, 1 "Villager" cards, 1 "Berry Bush" and 5 "Berry" cards
    // code: await this.thereShouldBeNStackOfNSCardsNSCardsNSAndNSCards(1, 1, "Farm", 1, "Villager", 1, "Berry Bush", 5, "Berry")
    // hint: Post_20221013_StoneHouse_Context.thereShouldBeNStacksOfNSCardsNSCardsAndNSCards

    var actual = expected; // FIXME
    expect(actual).toEqual(expected);

  }

  async afterTest() {
    // Do your teardown here, if necessary
  }
}

File: src/test/java/com/drpicox/game/building/Post_20221019_Farm_Context.java
package com.drpicox.game.building;

import org.springframework.stereotype.Component;
import static com.google.common.truth.Truth.assertThat;
import static com.google.common.truth.Truth8.assertThat;
import com.drpicox.game.util.FrontendSimulator;
import com.drpicox.game.game.GivenGameService;
import com.drpicox.game.card.GivenCardService;
import com.drpicox.game.game.api.GameDTO;

@Component
public class Post_20221019_Farm_Context {

    private final FrontendSimulator frontendSimulator;
    private final GivenGameService givenGameService;
    private final GivenCardService givenCardService;
    private GameDTO gameDTO;

    Post_20221019_Farm_Context(FrontendSimulator frontendSimulator, GivenGameService givenGameService, GivenCardService givenCardService) {
        this.frontendSimulator = frontendSimulator;
        this.givenGameService = givenGameService;
        this.givenCardService = givenCardService;
    }

    public void beforeTest() throws Throwable {
        // Do your setup here
        givenGameService.givenGame("empty");
        givenCardService.givenCards(1, "Berry");
        gameDTO = frontendSimulator.get("/api/v1/game", GameDTO.class);
        
        // Please, verify that:
        // [ ] there are villagers, militia, ... that need berries? How many? How many moons?
        // [ ] is the empty game right for this post?
        throw new UnsupportedOperationException("Please, review the implementation of beforeTest() and remove this exception when it is correct.");
    }

    public void thereShouldBeNSCard(int expected, String s1) {
        // text:  * there should be 0 "Farm" card.
        // code: this.thereShouldBeNSCard(0, "Farm")
        // hint: Post_20221020_WolfCallsWolf_Context.thereShouldBeNSCards

        var actual = expected; // FIXME
        assertThat(actual).isEqualTo(expected);


        throw new UnsupportedOperationException("The method thereShouldBeNSCard(int expected, String s1) is not implemented yet.");
    }

    public void givenNSNSNSCards(int n1, String s1, int n2, String s2, int n3, String s3) {
        // text:  * given 1 "Build Idea", 1 "Stone", 2 "Wood" cards
        // code: this.givenNSNSNSCards(1, "Build Idea", 1, "Stone", 2, "Wood")

        // Add here what is given

        // And make sure that the game is in the right state (also for the frontend)
        gameDTO = frontendSimulator.get("/api/v1/game", GameDTO.class);

        throw new UnsupportedOperationException("The method givenNSNSNSCards(int n1, String s1, int n2, String s2, int n3, String s3) is not implemented yet.");
    }

    public void givenThereIsTheSCardAtLevelNAndNXp(String s1, int n1, int n2) {
        // text:  * Given there is the "Build Idea" card at level 2 and 6 xp
        // code: this.givenThereIsTheSCardAtLevelNAndNXp("Build Idea", 2, 6)

        // Add here what is given

        // And make sure that the game is in the right state (also for the frontend)
        gameDTO = frontendSimulator.get("/api/v1/game", GameDTO.class);

        throw new UnsupportedOperationException("The method givenThereIsTheSCardAtLevelNAndNXp(String s1, int n1, int n2) is not implemented yet.");
    }

    public void endTheCurrentMoon() {
        // text:  * end the current moon.
        // code: this.endTheCurrentMoon()
        // hint: Post_20221024_Pickaxe_Context.endTheCurrentMoon

        gameDTO = frontendSimulator.post("/api/v1/game/moon", null, GameDTO.class);

        throw new UnsupportedOperationException("The method endTheCurrentMoon() is not implemented yet.");
    }

    public void thereShouldBeNSNSAndNSCards(int expected, String s1, int n2, String s2, int n3, String s3) {
        // text:  * there should be 0 "Stone", 0 "Wood" and 0 "Berry Bush" cards.
        // code: this.thereShouldBeNSNSAndNSCards(0, "Stone", 0, "Wood", 0, "Berry Bush")

        var actual = expected; // FIXME
        assertThat(actual).isEqualTo(expected);


        throw new UnsupportedOperationException("The method thereShouldBeNSNSAndNSCards(int expected, String s1, int n2, String s2, int n3, String s3) is not implemented yet.");
    }

    public void givenANewEmptyGame() {
        // text:  * Given a new empty game
        // code: this.givenANewEmptyGame()

        // Add here what is given

        // And make sure that the game is in the right state (also for the frontend)
        gameDTO = frontendSimulator.get("/api/v1/game", GameDTO.class);

        throw new UnsupportedOperationException("The method givenANewEmptyGame() is not implemented yet.");
    }

    public void givenThereAreNSCards(int n1, String s1) {
        // text:  * Given there are 1 "Farm" cards.
        // code: this.givenThereAreNSCards(1, "Farm")
        // hint: Post_20221020_WolfCallsWolf_Context.givenThereAreNSCards

        // Add here what is given

        // And make sure that the game is in the right state (also for the frontend)
        gameDTO = frontendSimulator.get("/api/v1/game", GameDTO.class);

        throw new UnsupportedOperationException("The method givenThereAreNSCards(int n1, String s1) is not implemented yet.");
    }

    public void givenThereAreNSCardsOrAnotherPlant(int n1, String s1) {
        // text:  * Given there are 1 "Berry Bush" cards or another plant.
        // code: this.givenThereAreNSCardsOrAnotherPlant(1, "Berry Bush")

        // Add here what is given

        // And make sure that the game is in the right state (also for the frontend)
        gameDTO = frontendSimulator.get("/api/v1/game", GameDTO.class);

        throw new UnsupportedOperationException("The method givenThereAreNSCardsOrAnotherPlant(int n1, String s1) is not implemented yet.");
    }

    public void thereShouldBeNSCards(int expected, String s1) {
        // text:  * There should be 5 "Berry" cards.
        // code: this.thereShouldBeNSCards(5, "Berry")
        // hint: Post_20221020_WolfCallsWolf_Context.thereShouldBeNSCards

        var actual = expected; // FIXME
        assertThat(actual).isEqualTo(expected);


        throw new UnsupportedOperationException("The method thereShouldBeNSCards(int expected, String s1) is not implemented yet.");
    }

    public void afterTest() {
        // Do your teardown here, if necessary
    }
}
